Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,245 @@
+        
+        let requiredSize = Posix.cmsgSpace(payloadSize: MemoryLayout.stride(ofValue: payload))
+        precondition(writableBuffer.count >= requiredSize, "Insufficient size for cmsghdr and data")
+        
+        let bufferBase = writableBuffer.baseAddress!
+        // Binding to cmsghdr is safe here as this is the only place where we bind to non-Raw.
+        let cmsghdrPtr = bufferBase.bindMemory(to: cmsghdr.self, capacity: 1)
+        cmsghdrPtr.pointee.cmsg_level = level
+        cmsghdrPtr.pointee.cmsg_type = type
+        cmsghdrPtr.pointee.cmsg_len = .init(Posix.cmsgLen(payloadSize: MemoryLayout.size(ofValue: payload)))
+        
+        let dataPointer = Posix.cmsgData(for: cmsghdrPtr)
+        precondition(dataPointer!.count >= MemoryLayout<PayloadType>.stride)
+        dataPointer!.storeBytes(of: payload, as: PayloadType.self)
+        
+        self.writePosition += requiredSize
+    }
+    
+    /// The result is only valid while this is valid.
+    var validControlBytes: UnsafeMutableRawBufferPointer {
+        if writePosition == 0 {
+            return UnsafeMutableRawBufferPointer(start: nil, count: 0)
+        }
+        return UnsafeMutableRawBufferPointer(rebasing: self.controlBytes[0 ..< self.writePosition])
+    }
+    
+}
+
+extension UnsafeOutboundControlBytes {
+    /// Add a message describing the explicit congestion state if requested in metadata and valid for this protocol.
+    ///  Parameters:
+    ///     - metadata:   Metadata from the addressed envelope which will describe any desired state.
+    ///     - protocolFamily:  The type of protocol to encode for.
+    internal mutating func appendExplicitCongestionState(metadata: AddressedEnvelope<ByteBuffer>.Metadata?,
+                                                         protocolFamily: NIOBSDSocket.ProtocolFamily?) {
+        guard let metadata = metadata else { return }
+
+        switch protocolFamily {
+        case .some(.inet):
+            self.appendControlMessage(level: .init(IPPROTO_IP),
+                                      type: IP_TOS,
+                                      payload: CInt(ecnValue: metadata.ecnState))
+        case .some(.inet6):
+            self.appendControlMessage(level: .init(IPPROTO_IPV6),
+                                      type: IPV6_TCLASS,
+                                      payload: CInt(ecnValue: metadata.ecnState))
+        default:
+            // Nothing to do - if we get here the user is probably making a mistake.
+            break
+        }
+    }
+}
+
+extension AddressedEnvelope.Metadata {
+    /// It's assumed the caller has checked that congestion information is required before calling.
+    internal init(from controlMessagesReceived: UnsafeControlMessageCollection) {
+        var controlMessageReceiver = ControlMessageParser()
+        controlMessagesReceived.forEach { controlMessage in controlMessageReceiver.receiveMessage(controlMessage) }
```

Comment:
```swift
for message in controlMessagesReceived {
    controlMessageReceiver.receiveMessage(message)
}
```

In general `forEach` is not really worth using: it's kinda needless sugar over the more straightforward looping operation.

Reply:
Also, should this loop move into an `init` on `ControlMessageParser`? That object can really only be used with one thing _anyway_.

Reply:
ok